<launch>
  <arg name="camera1_ns" default="camera_a/rgb" />
  <arg name="camera1_image" default="image_rect" />
  <arg name="camera2_ns" default="camera_b/rgb" />
  <arg name="camera2_image" default="image_rect" />
  <arg name="checker_rows" default="4"/>
  <arg name="checker_cols" default="5"/>
  <arg name="checker_size" default="0.0245"/>
  <arg name="headless" default="false" />

  <!-- checkerboard detector for each camera -->
  <include file="$(find camera_pose_calibration)/blocks/rgb_block.launch">
    <arg name="ns" value="$(arg camera1_ns)" />
    <arg name="image_topic" value="$(arg camera1_image)" />
    <arg name="checker_rows" value="$(arg checker_rows)" />
    <arg name="checker_cols" value="$(arg checker_cols)" />
    <arg name="checker_size" value="$(arg checker_size)" />
  </include>
  <include file="$(find camera_pose_calibration)/blocks/rgb_block.launch">
    <arg name="ns" value="$(arg camera2_ns)" />
    <arg name="image_topic" value="$(arg camera2_image)" />
    <arg name="checker_rows" value="$(arg checker_rows)" />
    <arg name="checker_cols" value="$(arg checker_cols)" />
    <arg name="checker_size" value="$(arg checker_size)" />
  </include>

  
  <!-- find time intervals where all camera's see a checkerboard -->  
  <node type="interval_intersection_action" pkg="interval_intersection" name="interval_intersection" output="screen" />
  <node pkg="camera_pose_calibration" type="start_interval_intersection.py" name="start_interval_intersection"
        args="$(arg camera1_ns) $(arg camera2_ns)" output="screen" />
  <node pkg="camera_pose_calibration" type="filter_intervals.py" name="filter_intervals">
     <param name="min_duration" value="1.0" />
     <remap from="features" to="$(arg camera1_ns)/features" />
     <remap from="interval" to="intersected_interval"/>
  </node>


  <!-- generate robot measurements -->
  <node pkg="camera_pose_calibration" type="multicam_capture_exec.py" name="capture_exec" 
        args="$(arg camera1_ns) $(arg camera2_ns)" output="screen">
    <param name="cam_info_topic" value="camera_info" />
    <remap from="request_interval" to="interval_filtered" />
  </node>


  <!-- optimize camera poses -->
  <node type="run_optimization_online.py" pkg="camera_pose_calibration" name="cal_optimizer" output="screen"/>


  <!-- Visualization -->
  <node type="capture_monitor.py" pkg="camera_pose_calibration" name="capture_monitor" output="screen"
        args="$(arg camera1_ns) $(arg camera2_ns)" />

  <node unless="$(arg headless)" type="image_view" pkg="image_view" name="aggregated_image_viewer" >
     <remap from="image" to="aggregated_image" />
  </node>

</launch>
